[Feat #26] JWT 관리(저장/재발급) 및 로그아웃 API

dudxo opened the pull request:
### 관련 이슈
- close #26 

### 📑 작업 상세 내용
- 로그인 시 발급 되는 RefreshToken 관리(Redis 저장)
- 로그아웃 API
  - Redis 저장된 RefreshToken 삭제
  - 발급되어 있는 AccessToken BlackList 지정
- AccessToken 재발급 API
  - 로그아웃 또는 만료된 Access/RefreshToken 검증
  - Access/Refresh Token 재발급
  - Redis Refresh 업데이트

### 💫 작업 요약
- 로그아웃 API
- 로그아웃 API 테스트
- 토큰 재발급 API
- 토큰 재발급 API 테스트

### 🔍 중점적으로 리뷰 할 부분
- TokenProvider.class <-> MemberService.class에서 순환참조가 발생해서 임시방편으로 TokenProvider.class에서 MemberService가 아닌 MemberRepository를 의존하도록 변경했습니다.
- 로그아웃/토큰 재발급 API 단위 테스트를 거의 다 Mocking해서 테스트 했는데 이게 맞는가 싶네요,, 다 가짜로 주입하면 이게 성공할 수 밖에 없는 구조아닌가 싶어요,,

dudxo committed:
[fix] : TokenService 삭제

dudxo committed:
[feat] : 소셜 로그인 성공 시 토큰 생성 로직 수정
- refreshToken 생성 추가
- refreshToken 생성 시 redis 저장 추가

dudxo committed:
[feat] : Token 관련 에러코드 추가

dudxo committed:
[feat] : 남은 만료시간 반환 메서드 추가

dudxo committed:
[feat] : 로그아웃 로직 추가

dudxo committed:
[feat] : DTO Valid 추가

dudxo committed:
[fix] : logout HttpMethod 변경 및 RequestBody 추가

dudxo committed:
[feat] : accessToken logout 확인 로직 추가

dudxo committed:
[feat] : 토큰 재발급 로직 추가

dudxo committed:
[refactor] : 변경감지를 이용한 ResponseDTO 생성

dudxo committed:
[test] : Mocking을 이용한 Service 테스트 수정

dudxo committed:
[feat] : 로그아웃 실패 검증 로직 추가

dudxo committed:
[feat] : 로그아웃 에러코드 추가

dudxo committed:
[fix] : 토큰 재발급 반환 타입 DTO 오류 수정

dudxo committed:
[fix] : API '/' 누락 수정

dudxo committed:
[feat] : 토큰 재발급 시 생성을 위한 인증 객체 추가

dudxo committed:
[fix] : MeberService.class <-> TokenProvider.class 순환참조 해결

dudxo committed:
[test] : 로그아웃, 토큰 재발급 단위테스트 추가

dudxo committed:
Merge branch 'dev' into feat/26/JWT-Logout

dudxo commented:
헉 컨트롤러 통합 테스트를 작성 안했네요! 추가 작성해서 커밋하겠습니다,,

hyun2371 commented on src/main/java/com/dnd/gongmuin/member/dto/request/LogoutRequest.java:
```diff
@@ -0,0 +1,10 @@
+import jakarta.validation.constraints.NotEmpty;
+
+public record LogoutRequest(
+	@NotEmpty(message = "잘못된 요청입니다.")
```
accessToken을 입력해주세요 등의 메세지도 괜찮을 것 같아요!

hyun2371 commented on src/main/java/com/dnd/gongmuin/member/exception/MemberErrorCode.java:
```diff
@@ -11,7 +11,8 @@ public enum MemberErrorCode implements ErrorCode {
 	NOT_FOUND_MEMBER("특정 회원을 찾을 수 없습니다.", "MEMBER_001"),
 	NOT_FOUND_NEW_MEMBER("신규 회원이 아닙니다.", "MEMBER_002"),
-	NOT_ENOUGH_CREDIT("보유한 크레딧이 부족합니다.", "MEMBER_003");
+	FAIL_LOGOUT("로그아웃을 실패했습니다.", "MEMBER_003"),
```
로그아웃을 시도하고 실패한거니까 logout failed도 괜찮을것 같아요

dudxo committed:
[fix] : @AuthenticationPrincipal를 통해 획득한 인증 객체 타입 변경

dudxo committed:
[test] : setUp시 refreshToken 생성 및 저장 추가

dudxo committed:
[test] : MemberController 통합테스트 추가

dudxo committed:
[test] : 각 테스트 후 redis 초기화 추가

dudxo commented:
> 헉 컨트롤러 통합 테스트를 작성 안했네요! 추가 작성해서 커밋하겠습니다,,

>> 커밋 완료했습니다!

hyun2371 commented:
```java
if (!isOfficialEmail(findMember)) {
  new NotFoundException(MemberErrorCode.NOT_FOUND_NEW_MEMBER);
}
```
브랜치에서 풀받고 보니까 MemberService에 throw가 빠진 것 같아요!

hyun2371 reviewed (approved):
고생하셨습니다~

hyun2371 commented:
`로그아웃/토큰 재발급 API 단위 테스트를 거의 다 Mocking해서 테스트 했는데 이게 맞는가 싶네요,, 다 가짜로 주입하면 이게 성공할 수 밖에 없는 구조아닌가 싶어요,,`

서비스 내부 로직이 다 호출되는지만 확인하면 되지 않을까요? 통합 테스트에서 확인해도 괜찮을 것 같습니다!!

dudxo commented:
> `로그아웃/토큰 재발급 API 단위 테스트를 거의 다 Mocking해서 테스트 했는데 이게 맞는가 싶네요,, 다 가짜로 주입하면 이게 성공할 수 밖에 없는 구조아닌가 싶어요,,`
> 
> 서비스 내부 로직이 다 호출되는지만 확인하면 되지 않을까요? 통합 테스트에서 확인해도 괜찮을 것 같습니다!!

>> 그렇겠죠? 감사합니다 ㅎㅎ

dudxo committed:
[feat] : 로그아웃/토큰 재발급 요청 DTO Valid Message 수정

dudxo committed:
[feat] : 로그아웃 에러 코드 message 수정

dudxo committed:
[feat] : 로그아웃 에러 코드 message 수정

dudxo committed:
[style] : isOfficialEmailExists() 인라인 수정

dudxo committed:
[refactor] : 메일 본문 내용 상수 선언을 통한 리팩토링

dudxo committed:
[fix] : MailService <-> MemberService 순환 참조 위험성 제거를 위한 변경
- MailService가 MemberService가 아닌 MemberRepository를 의존하도록 변경
- 변경으로 인한 checkDuplicatedOfficialEmail() 로직 변경 및 ReadOnly 추가

dudxo committed:
[fix] : throw 누락 수정

dudxo committed:
[fix] : updateAdditionalInfo() 변경 감지에 의한 save() 로직 삭제

dudxo committed:
[style] : 반환 변수명 수정(findMember -> foundMember)

dudxo committed:
[refactor] : Enum 재사용성을 위한 parseProviderFromSocialEmail() 로직 리팩토링

dudxo committed:
[test] : TokenProviderTest 추가

dudxo committed:
[rename] : isDuplicatedNickname() 닉네임 존재 여부 변수명 변경

dudxo committed:
[test] : 실제 로직 변경으로 불필요해진 stub 제거

dudxo committed:
[test] : 회원가입 API 통합 테스트를 위한 신규 회원 저장 로직 추가

dudxo committed:
[test] : 실제 로직 변경으로 stub 변경 및 불필요한 mock 제거

dudxo committed:
[test]: MemberFixture 객체 생성 메서드 추가

hyun2371 commented on src/main/java/com/dnd/gongmuin/member/service/MemberService.java:
```diff
@@ -54,10 +55,10 @@ public Member saveOrUpdate(Oauth2Response oauth2Response) {
-		} else if (socialEmail.contains("NAVER")) {
+		} else if (socialEmail.contains(Provider.NAVER.getProvider())) {
 			return "NAVER";
 		}
```
```java
public static Provider fromSocialEmail(String socialEmail){
for (Provider provider : values()) {
  if (socialEmail.contains(provider.getProvider())) {
  return provider;
  }
}
throw new NotFoundException(AuthErrorCode.NOT_FOUND_PROVIDER);
}
```

enum 안에 함수를 넣어도 괜찮을 것 같아요!

dudxo replied:
책임 분배를 생각 못했네요! 반복문 + 조건식보다 스트림이 더 좋겠다고 생각되서 스트림 방식으로 변경했습니다~
추가적으로 Provider Test 코드도 작성했습니당

dudxo committed:
[refactor] : 소셜 이메일을 통한 Provider 찾는 로직 변경

dudxo committed:
[test] : parseProviderFromSocialEmail() 실제 로직 변경으로 인한 staticMock 추가

dudxo committed:
[test] : Provider Test 작성
